File: PracticeProblems/CPSolns/jumpGameII.cpp
#include<iostream>
#include<vector>
#include<limits.h>

using std::vector;
using std::cout;
using std::endl;
using std::min;

vector<int> store;

int helper(vector<int> arr, int start){
    if(start >= arr.size() - 1)
        return 0;
    if(store[start])
        return store[start];

    int minJumps = INT_MAX;
    for(int i = start + 1; i <= start + arr[start]; i++)
        minJumps = min(minJumps, 1 + helper(arr, i));
    
    return store[start] = minJumps;
}

int jump(vector<int> &arr){
    store = vector<int>(arr.size());
    return helper(arr, 0);
}

int main(){
    vector<int> arr = {2, 3, 1, 1, 4};
    cout << jump(arr) << endl;
    return 0;
}

File: PracticeProblems/JavaSolns/JumpGameII.java
//Leet code 45

class JumpGameII {
    private static Integer[] store;

    private static int helper(int[] A, int start) {
        if (start >= A.length - 1)
            return 0;
        if (store[start] != null)
            return store[start];

        int minJump = Integer.MAX_VALUE;
        for (int i = start + 1; i <= start + A[start]; i++)
            minJump = Math.min(minJump, 1 + helper(A, i));

        return store[start] = minJump;
    }

    public static int jump(int[] A) {
        store = new Integer[A.length];
        return helper(A, 0);
    }

    public static void main(String args[]){
        int[] arr = {2, 3, 1, 1, 4};
        System.out.println(jump(arr));
    }
}
